enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+
+#### Phase 2: Quality of Life (QoL) Improvements
+
+Summary: Prepare for general availability by tackling any performance improvements, shoring up the overall design/implementation/UX, and any QoL improvements that have been identified from previous phase feedback.
+
+This phase has largely been unexplored. Candidates for this phase may include:
+
+- Integration with the OCP console, hypershift, KCP, etc.
+- Improved instrumentation.
+- Improved performance of the overall stack.
+- Improved UX around the user-facing APIs.
+
+### PO Requirements
+
+- Must package Kubernetes manifests using OLM's registry+v1 bundle format.
+- Cannot specify non-AllNamespace CSV install modes.
+- Cannot declare package or GVK dependencies.
+- All package bundles must live in the redhat-operators CatalogSource.
+
+### High-Level Design
+
+In order to enable admins to express their desired intent to install an OLM-based operator in day 0 operations, a new, top-level `PlatformOperator` cluster-scoped CRD will be introduced. This resource can help serve as the bridge between the openshift installer, and the on-cluster components.
+
+The initial API design will include a spec.Package field, which references an OLM-based operators's package name:
+
+```yaml=
+apiVersion: platform.openshift.io/v1alpha1
+kind: PlatformOperators
+metadata:
+  name: cert-manager
+spec:
+  package: openshift-cert-manager-operator
+```
+
+By default, this resource will contain an empty spec. Future phase requirements may necessitate more granular admin controls over how upgrades are approved, and how PO packages are sourced. During phase 0, there may be a limitation on which catalog sources are supported.
+
+Before cluster creation, an admin would be responsible for creating, and editing, the cluster's Kubernetes manifests with their desired PlatformOperator(s) configuration, and a FeatureGate manifest that enables tech preview features. The installer would then bootstrap the cluster using those custom manifests.
+
+During cluster rollout, the CVO will deploy the PO manager component. The PO manager will watch the PlatformOperators custom resources, and ensure that the desired PO packages are managed on the cluster.
+
+At a high-level, this component is responsible for the following:
+
+- Sourcing OLM bundles from supported catalog sources.
+- Applying the sourced bundles, and persisting their manifests to the cluster.
+- Relaying underlying bundle state back to the CVO.
+
+The PO manager will manage a ClusterOperator resource, which will act as a proxy for communicating the aggregate health of the PO packages back to the CVO. Future phases may require PO packages to reconcile OLM's OperatorCondition resource, or directly manage a ClusterOperator resource so the PO manager can properly relay package state back to the CVO.
```

Comment:
> Am I understanding this correctly that the status of all the PlatformOperators, even if there are many, would be aggregated to a single ClusterOperator resource

Yes.  This is effectively a constraint of the CVO.  The CVO only "cares about" CO resources that it manages (i.e. are part of the payload).  Since POs are, by definition, not part of the payload, there's no way for their associated COs to be part of the payload, so the CVO won't care about them.  In order to make the CVO care about the status of the POs for purposes of determining when an install/upgrade is complete, the platform operator manager will need to proxy the PO status onto a CO that lives in the payload, which means a single static CO.

Each PO can certainly *also* manage its own CO for purposes of reporting information to the admin, but the CVO won't care about it.

> That will especially be interesting if operators that currently have status in a CO become optional and their CO goes away.

The CO doesn't have to go away, but the CVO won't care about it any more because it's not part of the payload.

Reply:
> Is there a reason not to make the CVO care about all CO resources?

we may get there, right now we wanted a path that would let us make progress w/o being blocked by the CVO team agreeing to and making such a change.  I don't recall if explicit reasons have been presented for why it would be undesirable or technically problematic for the CVO to do so.

Reply:
Is there anything we can do in the immediate term outside of highlighting this behavior in this enhancement? We've discussed this before, but requiring OLM-based operators to reconcile a ClusterOperator resource seems like a difficult sell, and it's unclear whether OLM's existing OperatorConditions API will stick around. It would be nice to iron out this interaction with the CVO during phase 1.

Reply:
I would prefer not to tie POs to the CO api because, fundamentally, they aren't COs and much as we want them to largely act like POs, there are probably enough differences to justify not reusing that api.

so personally i'm comfortable with:

1) POM proxies/aggregates to a CO for CVO for now
2) assuming we want POs to directly report/manage their conditions somewhere (i think we will, someday), that should be either a common resource defined by OLM (e.g. OperatorConditions), or a new PO-specific status resource
3) Whether we'd ever want the CVO to directly respect the status being reported by an operator (instead of relying on the POM proxy).......maybe?  but doesn't strike me as our most pressing concern.

Reply:
Okay, that seems reasonable to me, especially for this current phase. I already have "Require PO packages to reconcile a ClusterOperator/OperatorCondition resource during phase 0." documented as a non-goal for this proposal, which is something we can revisit for phase 1. Is there anything else we'd need to do in the short term here?

Reply:
Marking this as resolved after updated the proposal with the CVO limitation and the potential want/need for an auxiliary API for POs to communicate their overall health vs. managing an "aggregate" CO to relay state during phase 0.